opt: explain execution plans

RaduBerinde opened the pull request:
Adding support for returning the EXPLAIN of a generated execution
plan. This can be used with the `exec-explain` test directive.

Also fixing the planner setup to use an internal planner.

Release note: None

RaduBerinde committed:
opt: explain execution plans

Adding support for returning the EXPLAIN of a generated execution
plan. This can be used with the `exec-explain` test directive.

Also fixing the planner setup to use an internal planner.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21397)
<!-- Reviewable:end -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 6 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21397#-:-L2_fotX8HbZ2Bwv3TBh:bnfp4nl)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: clean up explainer flags

The previous commit tickled `go vet` and it now complains about
`explainer := explainer{..}`.
Storing the flags directly in explainer is a bit confusing, so I took
this opportunity to separate the flags into a separate structure; the
runtime state stays in `explainer`.

Release note: None

RaduBerinde commented:
I added a commit that cleans up some stuff in explain_plan. @knz can you take a look? Do you think that `expanded`, `optimized` belong in `explainFlags` as well?

#### sql: clean up explainer flags

The previous commit tickled `go vet` and it now complains about
`explainer := explainer{..}`.
Storing the flags directly in explainer is a bit confusing, so I took
this opportunity to separate the flags into a separate structure; the
runtime state stays in `explainer`.

Release note: None

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> - I think they belong for now.

The two explain options "expanded" and "optimized" have been rarely used in testing, and never outside of testing. I suppose we can let them fall into obsolescence, so don't sweat about them. With the new code, we'll have a different way to monitor how optimizations are being performed and I expect that the EXPLAIN interface will change radically to provide insight into that.

---

Reviewed 5 of 6 files at r1, 3 of 3 files at r2.
Review status: 5 of 6 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21397#-:-L2aEFP6oq0etridnnc0:brh319)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Thank you!
